### DISCUSSION TITLE
var, let, const의 차이점은 무엇인가요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#19] [변수]
- var, let, const의 차이점은 무엇인가요?
  - const는 재할당이 불가능한데, const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?

**Comment by pipisebastian** [comment]
>var, let, const의 차이점은 무엇인가요?

1. `var` 
    - 함수 레벨 스코프 ⇒ 함수 외부의 변수는 전역 변수가 됨 ⇒ **중복 선언 발생!**
    - 같은 스코프 내, **중복 선언 가능** ⇒ 의도치 않게 변수 재할당됨! ⇒ **의도치 않은 동작**
2. `let`
    - 블록 레벨 스코프, **중복 선언 불가능**, 값 재할당 **가능**
3.  `const`
    - 블록 레벨 스코프, **중복 선언 불가능**, 값 재할당 **불가능**(선언과 동시에 초기화)
    
>const는 재할당이 불가능한데, const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?

- 객체는 기본적으로 원시타입과 다르게 값이 아닌, 메모리주소를 저장합니다.
- 이 메모리 주소는 또 다시 객체가 담긴 값들의 주소를 저장하는데요, 그렇기 때문에 객체의 값을 수정한다는 것은, 실제로는 한단계를 더 거쳐서 값이 변하는 거라, 1차적인 변수가 저장하는 메모리 주소는 변하지 않습니다.

**Comment by publdaze** [comment]
> - var, let, const의 차이점은 무엇인가요?
- var은 함수 스코프, 재할당 가능
- let은 블록 스코프, 재할당 가능
- const는 블록 스코프, 선언과 동시에 초기화, 재할당 불가
> - const는 재할당이 불가능한데, const로 선언한 객체를 수정할 수 있는 이유는 무엇인가요?
- const 객체 변수는 값이 아닌 값이 저장되는 메모리힙 주소가 콜스택에 할당 됨. 그래서 객체 자체를 변경하는 게 아닌 개체 내부 프로퍼티를 변경할 경우에는 변수에 값을 재할당 한게 아닌 메모리 힙에 저장된 값이 변경되는 것으로 실제로 콜스택에 저장된 메모리힙 주소는 변경된 부분이 없음
